flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/CommittableToTableChangeConverter.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    Table table = tableLoader.loadTable();
+    this.io = table.io();
+    this.specs = table.specs();
+    this.tableName = table.name();
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext context) throws Exception {
+    manifestFilesCommitedState.clear();
+    manifestFilesCommitedState.addAll(manifestFilesCommitedList);
+  }
+
+  @Override
+  public void processElement(
+      CommittableMessage<IcebergCommittable> value,
+      ProcessFunction<CommittableMessage<IcebergCommittable>, TableChange>.Context ctx,
+      Collector<TableChange> out)
+      throws Exception {
+    if (value instanceof CommittableWithLineage) {
+      IcebergCommittable committable =
+          ((CommittableWithLineage<IcebergCommittable>) value).getCommittable();
+
+      if (committable == null || committable.manifest().length == 0) {
+        return;
+      }
+
+      DeltaManifests deltaManifests =
+          SimpleVersionedSerialization.readVersionAndDeSerialize(
+              DeltaManifestsSerializer.INSTANCE, committable.manifest());
+
+      List<DataFile> dataFiles = Lists.newArrayList();
+      List<DeleteFile> deleteFiles = Lists.newArrayList();
+
+      ManifestFile dataManifest = deltaManifests.dataManifest();
+      if (dataManifest != null) {
+        if (!manifestFilesCommitedSet.contains(dataManifest.path())) {
+          dataFiles.addAll(FlinkManifestUtil.readDataFiles(dataManifest, io, specs));
+          addManifestInfo(committable.checkpointId(), dataManifest.path());
+        } else {
+          LOG.info("Data Manifest file {} has already been committed", dataManifest.path());
+        }
+      }
+
+      ManifestFile deleteManifest = deltaManifests.deleteManifest();
+      if (deleteManifest != null) {
+        if (!manifestFilesCommitedSet.contains(deleteManifest.path())) {
+          try (CloseableIterable<DeleteFile> deleteFileIterable =
+              ManifestFiles.readDeleteManifest(deleteManifest, io, specs)) {
+            Iterables.addAll(deleteFiles, deleteFileIterable);
+          }
+
+          addManifestInfo(committable.checkpointId(), deleteManifest.path());
+        } else {
+          LOG.info("Delete manifest file {} has already been committed", deleteManifest.path());
+        }
+      }
+
+      if (dataFiles.isEmpty() && deleteFiles.isEmpty()) {
+        return;
+      }
+
+      TableChange tableChange = new TableChange(dataFiles, deleteFiles, false);
+      out.collect(tableChange);
+    } else {
+      LOG.warn("Unsupported type of committable message: {}", value.getClass());
+    }
+  }
+
+  private void addManifestInfo(long checkpointId, String manifestPath) {
+    manifestFilesCommitedSet.add(manifestPath);
+    manifestFilesCommitedList.add(Tuple2.of(checkpointId, manifestPath));
+
+    // If the capacity is exceeded, delete the earliest file.
+    if (manifestFilesCommitedSet.size() > maxSize) {
+      NavigableMap<Long, List<String>> commitRequestMap = Maps.newTreeMap();
```

Comment:
Why not store the `manifestFilesCommited` in this data structure immediately? Why do we calculate this again and again?